### PR TITLE
[WIP] Finalize Services

### LINKED ISSUE #2531: Optional attributes of serverconnection ISettings
A number of attributes at https://github.com/jupyterlab/jupyterlab/blob/34c666b0106b215c0c2cdbe6f4a2488be45af92a/packages/services/src/serverconnection.ts#L120 are declared as having a default, but are still not optional in the typings. Can we make the settings with defaults optional?

### LINKED ISSUE #3070: Clean up handling of file browser server errors
Right now we get a cryptic "Server Connection Error" when the file browser model can't complete an operation.  It isn't triggered uniformly for different server error types either.  We should clean this up and use an error message more like the one in the classic notebook:

```
A connection to the notebook server could not be established. The notebook will continue trying to reconnect. Check your network connection or notebook server configuration.
```

### LINKED ISSUE #3261: Request header field X-XSRFToken is not allowed by Access-Control-Allow-Headers in preflight response
I am using following tutorial to build sample jupyterlab extension
_http://jupyterlab-tutorial.readthedocs.io/en/latest/xkcd_extension_tutorial.html_
and run into run-time error on step 'Show a comic in the panel'
_lab:1 Failed to load https://egszlpbmle.execute-api.us-east-1.amazonaws.com/prod?1511122136445: Request header field X-XSRFToken is not allowed by Access-Control-Allow-Headers in preflight response._
my web client is windows chrome and the https request run fine if I manually run above url
I have attached zip with  jupyterlab_xkcd/src/index.ts 
[index.zip](https://github.com/jupyterlab/jupyterlab/files/1485950/index.zip)

Here is my local versions:
npm v 5.5.1
nodejs v 8.9.1
jupyterlab v 0.29.2
Could you advise what I might be doing wrong or suggest work around please
Thanks
Alex

### PR DESCRIPTION
**Pull request by blink1073**
Prepping for a 1.0 release of `services`.  Most of the line changes are from refactoring the tests.

Fixes #2531.  Fixes #3261. Fixes #3070.

- [x] Switch to using the `fetch` API instead of `XMLHttpRequest`.
- [x] Assert that `ServerConnection` only be used to talk to the notebook server
- [x] Add handling of config from the classic notebook (from `body` data)
- [x] Add `shutdownAll` methods to the managers.
- [x] Normalize the handling of server settings across the classes
- [x] Use the notebook server for tests for most of the suite
- [x] Add the `fetch` polyfill to the examples and the main app
- [x] Wrap a `TypeError` from a network failure into a `NetworkError`

### CODE REVIEW

**Comment by blink1073** [comment]
cc @lgeiger @yuvipanda

**Comment by blink1073** [comment]
@lgeiger, `index.js` in the node example is of most interest to you.

**Comment by vidartf** [comment]
> Assert that ServerConnection only be used to talk to the notebook server

In nbdime I'm using `ServerConnection` to connect to **either** a notebook server with a server extension, **or** a custom tornado server with handlers based on `notebook.base.handlers`. Will I still be able to use `ServerConnection`? What are the actual requirements for using it when using services as a library?

**Comment by lgeiger** [comment]
Thanks for the heads-up. Looking forward to 1.0

**Comment by blink1073** [comment]
@vidartf, are you calling urls that differ from the baseUrl, or are they always the same once you've chosen a server type?  Do all of those servers accept token/cookie based auth?

**Comment by blink1073** [comment]
Note that you can disable token auth by passing in an empty string (which we should document), I think if we also added the ability to disable the cookie auth then it would be fair to say that a given instance of a service should only communicate with one URL.

**Comment by blink1073** [comment]
@vidartf, @jasongrout, I believe https://github.com/jupyterlab/jupyterlab/pull/3294/commits/b45793afab9f0b6832202e4eb3bc36e3621b1863 addresses your respective concerns.

**Comment by vidartf** [comment]
@blink1073 The called URLs should be the same relative to the base URL. When *not* using the serverextension, authentication is disabled (no use of `@authenticated`). If at some point we decide to add that, it would be based on existing authenticator infrastructure (i.e. token/cookie).

That said, I was hoping for a more 'formalized' description of which features it assumes are available. Or rather, what the reason was for declaring that it should only be for ~notebook server (API dependency, security, ...).

**Comment by blink1073** [comment]
This sound good?

```
/**
 * The namespace for ServerConnection functions.
 *
 * #### Notes
 * This is only intended to enable communication with a server providing
 * Jupyter services, which has unique security requirements.
 * Unless a an empty string is given for `token`, token based authorization
 * is used.  If there is no token, XSRF cookies are used unless
 * `disableXSRF` is set to `true`.  A content type of `'application/json'` is
 * added if there is no body data to prevent malicious forms.
 *
 * Use the `fetch` API for requests to other servers.
 * https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
 */
```

**Comment by vidartf** [comment]
> This is only intended to enable communication with a server providing Jupyter services, which has unique security requirements.

I guess my confusion is whether this means talking to anything that is part of the [Jupyter services spec](https://github.com/jupyter/notebook/blob/master/notebook/services/api/api.yaml). Or if it means anything that inherits [AuthenticatedHandler](https://github.com/jupyter/notebook/blob/master/notebook/base/handlers.py#L59) (or follows its interface).

**Comment by blink1073** [comment]
How about this?

```
/**
 * The namespace for ServerConnection functions.
 *
 * #### Notes
 * This is only intended to enable communication with a server providing
 * Jupyter [services](https://github.com/jupyter/notebook/blob/master/notebook/services/api/api.yaml).
 *
 * If the server is using not using an authenticated handler, e.g. [AuthenticatedHandler](https://github.com/jupyter/notebook/blob/master/notebook/base/handlers.py#L59), set `authenticatedHandler` to `false`.
 *
 * When using an `authenticatedHandler`, we use `token` based authenication,
 * either the `token`, or a XSRF cookie if the token is an empty string.
 * A content type of `'application/json'` is added if there is no body data 
 * to allow the server to prevent malicious forms.
 *
 * Use the `fetch` API for requests to other servers.
 * https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
 */
```

**Comment by vidartf** [comment]
Just from looking at the source, maybe something like this?

```
/**
 * The namespace for ServerConnection functions.
 *
 * #### Notes
 * The ServerConnection should only be used to connect to the server
 * which served the current page (normally the notebook server).
 * Otherwise you might risk [... appropriate security warning].
 *
 * [...]
 */
```

The main points being:
 - Only connect to the source of the page (does the baseUrl contain the domain? otherwise the comparison against it might be inefficient for ensuring this).
 - Make the security risk of incorrect use explicit for anyone considering to use it, but also be explicit for when it is safe to use it (and probably should be used to ensure an XSRF is added).

The XSFR token is a standard tornado feature, so as far as I understand it, as long as an XSFR token is available in the cookie it should be added to the request. The important bit is that the request goes to the correct server (i.e. the initial code was OK, just the comment/docstring was confusing).

**Comment by blink1073** [comment]
The baseUrl contains the domain.  The user could be using a different type of server (such as `node` or `go`).   Also, it could be a cross-domain connection to the server: https://github.com/jupyter/notebook/blob/master/notebook/notebookapp.py#L596.

**Comment by blink1073** [comment]
How about this?

```
/**
 * The namespace for ServerConnection functions.
 *
 * #### Notes
 * This is only intended to enable communication with a server providing
 * Jupyter [services](https://github.com/jupyter/notebook/blob/master/notebook/services/api/api.yaml).
 *
 * Since access to the server means access to running arbitrary code, it is
 * important to restrict access to the server.
 *
 * The default security settings assume the use of an [AuthenticatedHandler](https://github.com/jupyter/notebook/blob/master/notebook/base/handlers.py#L59).
 *
 * We use `token` authentication if available, falling back on an XSRF
 * cookie if one has been provided.
 *
 * A content type of `'application/json'` is added when using authentication
 * and there is no body data to allow the server to prevent malicious forms.
 *
 * Use the `fetch` API for requests to other servers.
 * https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
 */
```

**Comment by blink1073** [comment]
More explicit docs based on video discussion with @vidartf:

```javascript
/**
 * The namespace for ServerConnection functions.
 *
 * #### Notes
 * This is only intended for use within this package to configure the
 * connection to a server providing Jupyter [services](https://github.com/jupyter/notebook/blob/master/notebook/services/api/api.yaml).
 *
 * We use `token` authentication if available, falling back on an XSRF
 * cookie if one has been provided.
 *
 * A content type of `'application/json'` is added when using authentication
 * and there is no body data to allow the server to prevent malicious forms.
 */
```

**Comment by blink1073** [comment]
I think https://github.com/jupyterlab/jupyterlab/pull/3294/commits/a7bfd3c970020d837244356374c202330e0151d1 better captures what this can be used for:

```javascript
/**
 * The namespace for ServerConnection functions.
 *
 * #### Notes
 * This is only intended to manage communication with the Jupyter server.
 *
 * The default values can used in a JupyterLab or Jupyter Notebook context.
 *
 * We use `token` authentication if available, falling back on an XSRF
 * cookie if one has been provided on the `document`.
 *
 * A content type of `'application/json'` is added when using authentication
 * and there is no body data to allow the server to prevent malicious forms.
```

**Comment by blink1073** [comment]
@vidartf, I'm done making changes, mind taking a look at `PageConfig` and `ServerConnection`?

**Comment by blink1073** [comment]
One shovelful at a time...

File: packages/services/src/serverconnection.ts [hunk 1]
```diff
@@ -47,328 +107,137 @@ namespace ServerConnection {
...
+    if (url.indexOf(settings.baseUrl) !== 0) {
+      throw new Error('Can only be used for notebook server requests');
     }
 
-    headers = settings.requestHeaders;
-    for (let prop in headers) {
-      xhr.setRequestHeader(prop, headers[prop]);
-    }
+    let request = new settings.Request(url, init);
+    request = new settings.Request(request, settings.init);
```
Comment (vidartf) [inline packages/services/src/serverconnection.ts]: This is not a pattern I could find anywhere else, so I had to check the specs for what would happen. AFAICT, the settings in `init` will be overridden by `settings.init` here, and should (?) be equivalent to doing
```ts
let request = new settings.Request(url, {...init, ...settings.init});
```
Is this what was intended? I would maybe expect the individual request's init to take precedence. It should probably be commented/documented somehow.

File: packages/coreutils/src/pageconfig.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ namespace PageConfig {
   export
   function getOption(name: string): string {
     if (configData) {
-      return configData[name] || '';
+      return configData[name] || Private.getBodyData(name);
```
Comment (vidartf) [inline packages/coreutils/src/pageconfig.ts]: This behavior is documented on several downstream uses, but should probably be documented on this class or method as well.

**Comment by blink1073** [comment]
Thanks @vidartf!  Updated based on your review.

**Comment by vidartf** [comment]
👍

**Pull request merged**